exasol/analytics/audit/audit.py

```diff
@@ -23,3 +37,130 @@ def __init__(
+          Log are wrapped into one insert statement before and one after. The
+          actual modifying query in between remains unchanged.  The insert
+          statements before and after record the timestamp and, if the
+          ModifyQuery represents an INSERT operation, the number of rows in
+          the modified table.
+        """
+        for query in queries:
+            if not query.audit:
+                yield query.query_string
+            elif isinstance(query, AuditQuery):
+                yield self._insert(query)
+            elif isinstance(query, ModifyQuery):
+                yield from self._wrap(query)
+            else:
+                raise TypeError(
+                    f"Unexpected type {type(query)} of query {query.query_string}"
+                )
+
+    def _insert(self, query: AuditQuery) -> str:
+        insert_statement = (
+            InsertStatement(self._column_names, separator=",\n  ")
+            .add_constants(query.audit_fields)
+            .add_scalar_functions(BaseAuditColumns.values)
+        )
+
+        suffix = ""
+        if query.select_with_columns:
+            alias = TableNameImpl("SUB_QUERY")
+            subquery_columns = {
+                c.name.name: ColumnName(c.name.name, alias)
+                for c in query.select_with_columns.output_columns
+            }
+            insert_statement.add_references(subquery_columns)
+            suffix = (
+                "\nFROM VALUES (1) CROSS JOIN\n"
+                f"  ({query.query_string}) as {alias.fully_qualified}"
+            )
+
+        return (
+            f"INSERT INTO {self.name.fully_qualified} (\n"
+            f"  {insert_statement.columns}\n"
+            ") SELECT\n"
+            f"  {insert_statement.values}{suffix}"
+        )
+
+    def _wrap(self, query: ModifyQuery) -> Iterator[str]:
+        """
+        Wrap the specified ModifyQuery it into 2 queries recording the
+        state before and after the actual ModifyQuery.
+
+        The state includes timestamps and optionally the number of rows of the
+        modified table, in case the ModifyQuery indicates potential changes to the
+        number of rows.
+        """
+        if query.db_operation_type != DbOperationType.INSERT:
+            yield query.query_string
+        else:
+            yield from [
+                self._count_rows(query, "Begin"),
+                query.query_string,
+                self._count_rows(query, "End"),
+            ]
+
+    def _count_rows(self, query: ModifyQuery, event_name: str) -> str:
+        """
+        Create an SQL INSERT statement counting the rows of the table
+        modified by ModifyQuery `query` and populate columns in the Audit
+        Table marked with "+":
+
+        + TIMESTAMP: BaseAuditColumns.values
+        + SESSION_ID: BaseAuditColumns.values
+        - RUN_ID
+        + ROW_COUNT: subquery
+        + SPAN_NAME: query.db_operation_type: DbOperationType
+        - SPAN_ID
```

Comment:
I guess the LogSpan class you add later? Because without it the log entries for the Insert can't be associated easily with each other 

Also, I think, we called the columns LOG_SPAN_ID

Reply:
Please see class [BaseAuditColumns](https://github.com/exasol/advanced-analytics-framework/blob/main/exasol/analytics/audit/columns.py): 
```python
SPAN_ID = decimal_column("LOG_SPAN_ID", precision=32)
```
I removed redundant prefixes from the attribute names while still keeping them as column names.

Reply:
to be honest, it only adds confusion

Reply:
Then I will change all attribute names to be identical to the names of the resp. columns.

Reply:
See last push.

Reply:
I propose adding the LogSpan class in a separate PR / ticket
* #272